scan_engine.cc

```diff
@@ -2107,6 +2107,25 @@ void ultrascan_port_probe_update(UltraScanInfo *USI, HostScanStats *hss,
       adjust_timing = false;
     adjust_ping = false;
   }
+  /* Do not slow down if
+     1)  we are in --defeat-icmp-ratelimit mode
+     2)  the new state is closed or filtered
+     3)  this is a UDP scan
+     We don't want to adjust timing when we get ICMP response, as the host might
+     be ratelimiting them. E.g. the port is actually closed, but the host ratelimiting
+     ICMP responses so we had to retransmit the probe several times in order to
+     match the (slow) rate limit that the target is using for responses. We
+     do not want to waste time on such ports.
+     On the other hand if the port is detected to be open it is a good idea to
+     adjust timing as we could have done retransmissions due to conjested network */
+  if (rcvdtime != NULL
+      && o.defeat_icmp_ratelimit
+      && (newstate == PORT_CLOSED || newstate == PORT_FILTERED)
```

Comment:
Hello Vincent,

This change would not make any sense because `(newstate == PORT_OPEN && newstate == PORT_FILTERED)` will always be false, as new state cannot be both in `PORT_OPEN` and `PORT_FILTERED` states.

There is another option to change the condition as following:
`&& (newstate == PORT_CLOSED || (newstate == PORT_OPENFILTERED))`
but, as I have already pointed out, this will not work. By default if we use UDP scan, all ports are initialized with `PORT_OPENFILTERED` state. This makes sense because we do not know yet if the port is open/closed/filtered, we can draw our conclusion only when we get an ICMP message (probably meaning that the port is closed or rate limited) or a UDP response (port open) or a timeout (contested network, firewall, rogue server). 

Check [this code](https://github.com/sergeykhegay/nmap/blob/a8d048f529440539f7e249606a471d274b16a508/scan_engine_raw.cc#L1879-L2007) out to understand how a port's state is changed. Basically, when we get an ICMP response during a UDP scan, the port's state can be changed to either `PORT_CLOSED` or `PORT_FILTERED`. But those ICMP responses might be rate limited, thus do not represent adequate information about probe timings, hence we do not adjust timings in the `--defeat-icmp-ratelimit` mode.

Makes sense?